### PR TITLE
git-flow: Proper autocomplete installation for ZSH

### PR DESCRIPTION
**Pull request by seivan**
Autocomplete doesn't work out of the box and it lacks a caveat telling you what to do. 
That being said, the caveat, as mentioned at the top of `git-flow-completion.zsh` is outdated, as it wouldn't be optimal. 

In the end, we decided to remove the ZSH completion all together and rely on 
```ruby
  def caveats
     <<~EOS
       To install Zsh completions:
         brew install zsh-completions
     EOS
   end
```

If you're using a plugin manager, like say `zinit` this also works
```sh
zinit ice lucid blockf wait depth"1" as"completion"
zinit snippet https://github.com/zsh-users/zsh-completions/blob/master/src/_git-flow
```

~~Renaming  `git-flow-completion.zsh` as `_git-flow` and replaces `"#!zsh` with  `"#compdef git-flow"`.
This breaks backwards compatibility, for existing users that source it, but they should change anyway.
Fixes the review comments:~~ 
https://github.com/Homebrew/homebrew-core/pull/93187#pullrequestreview-853563740
https://github.com/Homebrew/homebrew-core/pull/93187#pullrequestreview-853576361

~~So on post install it copies `git-flow-completion.zsh` as `_git-flow` and prepends `"#compdef git-flow"` to it before symlinking to `HOMEBREW_PREFIX/"share/zsh/site-functions"`.
That should make autocomplete work out of the box as long as `/usr/local/share/zsh/site-functions` is in `fpath`.
Also following other standards like `_brew` or `_aws`.~~

~~Maintaining backwards compatibility by keeping `git-flow-completion.zsh` symlinked and intact.~~

Also, I wonder, does this formula really need bottles as it's just a shell script?

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? I

-----

### CODE REVIEW

File: Formula/git-flow.rb
Comment (SMillerDev) [inline Formula/git-flow.rb:47]: We can also do:
```suggestion
      zsh_completion.install "git-flow-completion.zsh" => "_git_flow"
```

But all of this seems like something that should be fixed upstream

File: Formula/git-flow.rb
Comment (carlocab) [inline Formula/git-flow.rb:55]: I think we can get rid of all of this but the `inreplace`, which you can do in the `install` method.

File: Formula/git-flow.rb
Comment (seivan) [reply Formula/git-flow.rb:47]: That wouldn't work as far as I tested.
My suspicion is because it is installing from the bottle, that step is ignored.
Also it would break backwards compatibility. 

Fixing this upstream would break it for existing users. 
Also the hassle of fixing it upstream means touching a repo that hasn't been updated for seven years and then fixing new bottles, and so on.
This is quick fix that will work for both existing users sourcing it, as well as new installations.

File: Formula/git-flow.rb
Comment (seivan) [reply Formula/git-flow.rb:55]: 1. The install method is ignored if it's using bottles.
2. The file needs to be called `"_git-flow"`
3. Renaming the file alone would break for existing users.

Comment (seivan) [reply Formula/git-flow.rb:55]: Removing the bottles, and keeping it in the source makes sense to me. 
But keep in mind, renaming the original file will break things for people who are currently sourcing it. It may be the wrong approach but the file itself mentions to do so, so can't blame users for it.

**Comment by seivan** [comment]
The alternative is switching to using https://github.com/zsh-users/zsh-completions/blob/master/src/_git-flow as a resource for just ZSH as it comes complete, but that would break backward compatibility.

**Comment by SMillerDev** [comment]
That's already included in ZSH completions, we shouldn't be adding that to other formula

**Comment by seivan** [comment]
>

**Comment by seivan** [comment]
@SMillerDev 
> That's already included in ZSH completions, we shouldn't be adding that to other formula

I think you misunderstand. 
I didn't say add all of them, I said use that repository as a resource to extract the zsh-completions. 

As of now, the auto-complete is broken. 
You're distributing a file, that needs to be sourced, without mentioning that it needs to be sourced, more so, it's moved to site-functions that should be using compdef to be loaded, the file doesn't do that.
There are existing users, that *might* be sourcing it, so deleting it for them is not great. 

My suggestion is that you test the approaches you mentioned in your code review, so you can see that it doesn't work.

**Comment by seivan** [comment]
Another alternative is to remove the `zsh_completion.install` portion completely, updated the bottles so they don't distribute them (since bottles ignore `install`) and I can go on with my day with

```sh
zinit ice lucid blockf wait depth"1" as"completion"
zinit snippet https://github.com/zsh-users/zsh-completions/blob/master/src/_git-flow
```

**Comment by SMillerDev** [comment]
> You're distributing a file, that needs to be sourced, without mentioning that it needs to be sourced

It's in the upstream documentation, Homebrew shouldn't document upstream behavior.

> it's moved to site-functions that's already under fpathbut it doesn't work without `#compdef _git-flow´ and being named so.

This is a good suggestion to make to https://github.com/bobthecow/git-flow-completion so that it gets fixed for everyone, not just Homebrew.

> There are existing users, that might be sourcing it, so deleting it for them is not great.

It's not great, but if we move in that direction it'll always break some day when we remove the symlink since nobody will know it changed unless they read this PR. It's better to break it directly and let people fix it now.

File: Formula/git-flow.rb
Comment (carlocab) [reply Formula/git-flow.rb:55]: The bottles will be rebuilt when this PR is merged.

**Comment by seivan** [comment]
> > You're distributing a file, that needs to be sourced, without mentioning that it needs to be sourced
> 
> It's in the upstream documentation, Homebrew shouldn't document upstream behavior.

So what's the caveat used for in other packages? 


> This is a good suggestion to make to https://github.com/bobthecow/git-flow-completion so that it gets fixed for everyone, not just Homebrew.

That hasn't been touched in seven years, and has loads of issues. 
 


> It's not great, but if we move in that direction it'll always break some day when we remove the symlink since nobody will know it changed unless they read this PR. It's better to break it directly and let people fix it now.

Fine if that's okay with you, then my suggestion is to remove them, completely. 
Would that work out better for you?

File: Formula/git-flow.rb
Comment (SMillerDev) [reply Formula/git-flow.rb:47]: >  My suspicion is because it is installing from the bottle, that step is ignored.

If you're installing from a bottle locally it'll never pick up local modifications because it'll use the bottle definition of the formula.

> Fixing this upstream would break it for existing users.

It would stop being non-standard and work as ZSH users expect it to, yes

> Also the hassle of fixing it upstream means touching a repo that hasn't been updated for seven years and then fixing new bottles, tags, and so on.

Making new bottles is not that much work.

> This is quick fix that will work for both existing users sourcing it, as well as new installations.

But it will force homebrew to ship a workaround for eternity, [which we don't want](https://docs.brew.sh/Formula-Cookbook#patches)

File: Formula/git-flow.rb
Comment (seivan) [reply Formula/git-flow.rb:55]: @carlocab Ok, that works for me, but it will break things for existing users, is that fine?

**Comment by SMillerDev** [comment]
> So what's the caveat used for in other packages?

To document quirks that are specific to the way Homebrew installs software.

> Fine if that's okay with you, then my suggestion is to remove them, completely.

Since there are completions in a default ZSH install, that seems fine to me

**Comment by seivan** [comment]
> Since there are completions in a default ZSH install, that seems fine to me

Eh, what? 
That's not true.

File: Formula/git-flow.rb
Comment (seivan) [reply Formula/git-flow.rb:55]: The bottles being rebuilt, will keep the same hash, or do I need to do something?

Comment (carlocab) [reply Formula/git-flow.rb:55]: It's easy to fix, doesn't happen silently, and happens in an interactive context so users won't really miss it. Seems fine. If you're that concerned you can still a `git-flow-completion.sh` symlink. And/or document the change in the caveats.

Comment (seivan) [reply Formula/git-flow.rb]: Nah, I think @SMillerDev makes a good point of `But it will force homebrew to ship a workaround for eternity,`
Lets rip the band aid. 

I pushed the changes suggested.

File: Formula/git-flow.rb
Comment (seivan) [reply Formula/git-flow.rb]: Ok, added the changes.

**Comment by seivan** [comment]
@carlocab 
Just to be clear, if this merges, and I do `$homebrew update`, and `$homebrew upgrade` I will get the new bottles with these changes? 

The only way I could test locally, was to comment out all the bottles or use `--build-from-source`
I assume I don't need to do that if this merges into master?

**Comment by carlocab** [comment]
> [I]f this merges, and I do `$homebrew update`, and `$homebrew upgrade` I will get the new bottles with these changes?

Not in the current state of your PR. If you want this then you need to bump the `revision`. To get the new bottles with these changes after merging your current changes, one would need to do `brew update && brew reinstall git-flow`.

> @SMillerDev I am still open to removing the ZSH completion installation, since getting them from ` https://github.com/zsh-users/zsh-completions/blob/master/src/_git-flow` seems more stable somehow, as that repo is maintained, and the other one hasn't been touched for seven years.
> 
> Is this something you're ok with @carlocab ?

Fine with this, but let's document that in the `caveats`.

**Comment by SMillerDev** [comment]
I'd say this is enough as the caveat:
> Completions for ZSH are included in `brew install zsh-completions`

I don't think brew should recommend manually copying a file that a formula can install

**Comment by seivan** [comment]
> I'd say this is enough as the caveat:
> 
> > Completions for ZSH are included in `brew install zsh-completions`
> 
> I don't think brew should recommend manually copying a file that a formula can install

Yeah but you're forcing a ton of files to a user that might not need them, I certainly don't need them. But it's your call. I added the changes.

File: Formula/git-flow.rb
Comment (carlocab) [inline Formula/git-flow.rb:6]:
```suggestion
  revision 1

  stable do
```

File: Formula/git-flow.rb
Comment (carlocab) [inline Formula/git-flow.rb:52]:
```suggestion
      To install Zsh completions:
        brew install zsh-completions
```
Let's make this a bit more consistent with other formulae.

**Review by carlocab — COMMENTED** [review]
Thanks; could you squash your commits together please?

**Comment by seivan** [comment]
@carlocab 
> Thanks; could you squash your commits together please?

Done, and thanks for the help.

**Review by carlocab — APPROVED** [review]
Thanks, @seivan. Just pushed one small change to make sure the bottles are identical across build prefixes.

**Review by SMillerDev — APPROVED** [review]
Thanks @seivan  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Pull request closed without merging**
